service/storage_service.cc

```diff
@@ -681,7 +677,9 @@ void storage_service::join_token_ring(int delay) {
             // We might have not due to misconfiguration, e.g. skipping the "wait for gossip to settle" phase,
             // or due to a network partition which didn't allow us to contact other seeds.
             // But CDC requires that we know other nodes' tokens - they are needed to make a new generation of streams.
-            if (!wait_for_other_tokens(_token_metadata, _gossiper.get_seeds(), get_broadcast_address(), _abort_source)) {
+            // This is a heuristic: it assumes that if we managed to retrieve the tokens of all seeds,
+            // then we also managed to retrieve all other nodes' tokens.
+            if (!wait_for_seed_tokens(_token_metadata, _gossiper.get_seeds(), _abort_source)) {
```

Comment:
The wait_for_seed_tokens check is in the branch `!db::system_keyspace::bootstrap_complete()`, this means this is a bootstrapping node or a replacing node. The node should have learned the tokens before bootstrap or replacing, otherwise it will stream the wrong data.  We call `_gossiper.wait_for_range_setup().get()`  before and after the bootstrap node announces its token. We do similar thing for replace operation in the making replace node take writes [series](https://github.com/scylladb/scylla/pull/5731. ).

Reply:
No, it's not bootstrapping, that's the problem. It's skipping the streaming phase and joining the token ring immediately (it's either a seed or has auto_bootstrap=off). Otherise `_cdc_streams_ts` wouldn't be `nullopt`.

The bootstrapping case is handled in `bootstrap()`. There, `make_new_cdc_generation` is called after `wait_for_range_setup()`, as you said.

Reply:
Yes, this node can skip streaming (be a seed or auto_bootstrap = false), but it does not change the fact it is a bootstrapping node or replacing node, it is a new node joins the cluster because only new node will have bootstrap_complete() == false.

New node know should let other nodes know its token and learn other nodes token before it joins the cluster.

Reply:
> but it does not change the fact it is a bootstrapping node

Then I'm just using a different nomenclature. I was convinced that
- joining with streaming is called "bootstrapping"
- joining without streaming is not bootstrapping

Some parts of the code suggest that naming. Streaming is done inside the `bootstrap()` function, and if we join without streaming we skip the `bootstrap()` function.
Also, the "auto_bootstrap" flag suggest that bootstrapping = joining with streaming, because if we set it to `off`, the node will join without streaming.
But then there's the `bootstrap_complete()` function which says whether we have joined, not whether we have joined with streaming. IMO it should be called `joining_complete()`.
Inconsistent naming.

> New node know should let other nodes know its token

Maybe, but until now it wouldn't do that in the joining-without-streaming case. I'm not trying to change that in this PR, I'm just trying to make CDC work.

But I don't want to add another 30 second wait to the code. Also, relying on real time is not a correct solution anyway.